### ISSUE TITLE
Invalid links to key encoders in default provider

### ISSUE THREAD

**Issue by Lipovlan** [openssl/openssl#20738]
On [default provider manpage](https://www.openssl.org/docs/man3.1/man7/OSSL_PROVIDER-default.html) the links in section `Asymmetric Key Encoder` to [OSSL_ENCODER_RSA(7)](https://www.openssl.org/docs/man3.1/man7/OSSL_ENCODER-RSA.html) and similar are all invalid links.

**Comment by preveen-stack** [comment]
would like to attempt this. can anybody provide guidance. Looks like the referenced man pages are not yet written.  ref:  https://github.com/openssl/openssl/tree/master/doc/man7.

**Comment by mattcaswell** [comment]
Right, the man pages do not exist and need to be written. Incidentally, although this issue is specifically about the encoders, it looks like the decoders also do not exist, and in fact do not even have a section from them on the default provider manpage.

Each algorithm (RSA, DH, DSA, etc) needs a specific page written for it describing the specifics of the encoder for that algorithm including the parameters it understands along with (preferably) an example. There are no encoder pages at all but here's an example for the equivalent key management page:

https://www.openssl.org/docs/man3.1/man7/EVP_KEYMGMT-DH.html

The encoders themselves are implemented in providers/implementations/encode_decode in the various encode_*.c files.

There is some generic documentation on encoders in providers here:

https://www.openssl.org/docs/manmaster/man7/provider-encoder.html

You should also see the documentation on the various OSSL_ENCODER_* functions documented here:

https://www.openssl.org/docs/manmaster/man3/

Not sure I agree with the "good first issue" label on this. Getting a good enough understanding of the encoder framework in order to be able to write this documentation is non-trivial and quite a bit of work. But great if you want to do it. I've swapped the label to "help wanted" instead.

**Comment by preveen-stack** [comment]
This  [PR](https://github.com/openssl/openssl/pull/18969) was closed by the system as the author did not respond within 90 days. One of the changes requested was documentation of the implemented pubic API. Is there any general guidelines how the APIs are documented. Appreciate any pointers.

**Comment by mattcaswell** [comment]
All the documentation takes the form of .pod files in the man7/man5/man3/man1 sub-directories here:

https://github.com/openssl/openssl/tree/master/doc

man3 is for documentation of API functions, while man7 is more for overview type stuff. The documentation needed here is all man7. I'd suggest you take a look at the other man7 .pod files to get a general feel for how they are put together. If you add a new .pod file make sure you run "make update".

I'd recommend that you regularly run "make doc-nits" (having ensured "make update" was previously run if you have added new pod files) when you are writing the documentation. That complains about any documentation errors and looks for style issues. "make doc-nits" also gets run automatically by our CIs so a pull request adding documentation must have a "clean" output from that command.

**Comment by preveen-stack** [comment]
Ref: doc/HOWTO/documenting-functions-and-macros.md

**Comment by kovan** [comment]
This is addressed by PR #30297, which adds 9 new encoder/decoder documentation pages and updates the provider pages with cross-reference links. The same PR also addresses #17789 and #19641.
